### PR TITLE
Datasource Selection Documentation

### PR DESCRIPTION
**Pull request by JenMadiedo**
Issue Number: (Link to Github Issue or Azure Dev Ops Task/Story)

# DEPENDENT PR
> **Warning**
> This PR depends on commits in other PRs
> **DO NOT** merge this PR until https://github.com/Green-Software-Foundation/carbon-aware-sdk/pull/203 is merged and this branch is updated against the latest `dev` branch
## Summary
Add a doc that details the different configs/access patterns of the various data sources.

## Changes

- Add doc

## Checklist

- [ ] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [X] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

This PR Closes #<issue_number>

### CODE REVIEW

File: docs/data-source-matrix.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Data Source Matrices
```
Comment (bderusha) [inline docs/data-source-matrix.md:1]: Consider a title that tells the reader about the content rather than the format.

File: docs/data-source-matrix.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+- [Type of Data Sources and Configuration](#type-of-data-sources-and-configuration)
+- [Data Source Routes Available](#data-source-routes-available)
+- [Location Source Availability](#location-source-availability)
+
+## Type of Data Sources and Configuration
+
+In the CarbonAware SDK configuration, you can set what data source to use as the `EmissionsDataSource` and the `ForecastDataSource`. There are also certain configuration fields that must be set in order to access the raw data.
+| Type      | WattTime  | ElectricityMaps | JSON |
+|--------------------------|-----------|-----------------|------|
+| Is Emissions DataSource   | Yes      | Yes             | Yes     |
```
Comment (bderusha) [inline docs/data-source-matrix.md:16]: This is a great use-case for emojis.  Consider something like: ✅ ❌

File: docs/data-source-matrix.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+| Is Forecast DataSource   | Yes  | Yes            | No     |
+| Makes HTTP(s) call   | Yes  | Yes            | No    |
+| Can Use Custom Data   | No  | No            | Yes    |
+| Needs Authentication Config   | Yes - username/password | Yes - token/header            | No    |
```
Comment (bderusha) [inline docs/data-source-matrix.md:20]: For the purpose of this comparison, does the user need to be aware of the type of auth?  If no, consider removing.  If yes, consider referencing the nuance as a footnote.

Same comment applies to this entire matrix

File: docs/data-source-matrix.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+| BaseUrl Override Config   | Only for testing  | Switch between trial + full version, testing            | No   |
+| Supports Trial + Full Account   | Yes  | Yes (*different URL and token header required)            | N/A    |
+
+## Data Source Routes Available
```
Comment (bderusha) [inline docs/data-source-matrix.md:24]: I found this section pretty hard to parse.  Consider removing the list and imagining the entire thing as a table.  Potentially:

| Data Source Method | WattTime  | ElectricityMaps | JSON | CLI Usage | WebAPI Usage | SDK Usage|
|--------------|:-----------:|:-----------------:|:------:|:------:|:------:|:------:|
| `GetCarbonIntensityAsync` | ✅   | ✅ | ✅ | `emissions` | `emissions/bylocation` / `emissions/bylocations` / `emissions/bylocations/best` / `emissions/average-carbon-intensity` / `average-carbon-intensity/batch` | `GetEmissionsDataAsync(...)` / `GetBestEmissionsDataAsync(...)` / `GetAverageCarbonIntensityDataAsync(...)` |

File: docs/data-source-matrix.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+- GetCurrentForecastAsync
+  - CLI: `emissions-forecasts`
+  - API: `forecasts/current`
+  - Library: `GetCurrentForecastAsync(...)`
+- GetForecastByDateAsync
+  - CLI: `emissions-forecasts --requested-at`
+  - API: `forecasts/batch` with `requestedAt` field
+  - Library: `GetForecastByDateAsync(...)`
+
+| Route      | WattTime  | ElectricityMaps | JSON |
```
Comment (bderusha) [inline docs/data-source-matrix.md:41]: Consider calling these by their relationship to the data source: `Methods` instead of `Routes`  (here and in the section title)

File: docs/data-source-matrix.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+|--------------|:-----------:|:-----------------:|:------:|
+| GetCarbonIntensityAsync | Yes  | Yes            | Yes    |
+| GetCurrentForecastAsync | Yes  | Yes            | No    |
+| GetForecastByDateAsync | Yes  | No            | No     |
+
+## Location Source Availability
+
+*I wasn't sure exactly what to fill here but wanted something along the lines of:
```
Comment (bderusha) [inline docs/data-source-matrix.md:49]: Consider describing the fact that different data sources provide both different features (as outlined above) but also coverage of different geographic areas, then refer people to the most actionable place on their websites for details around which locations they support (or the API docs they would need to determine coverage)

File: docs/data-source-matrix.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  - API: `forecasts/batch` with `requestedAt` field
+  - Library: `GetForecastByDateAsync(...)`
+
+| Route      | WattTime  | ElectricityMaps | JSON |
+|--------------|:-----------:|:-----------------:|:------:|
+| GetCarbonIntensityAsync | Yes  | Yes            | Yes    |
+| GetCurrentForecastAsync | Yes  | Yes            | No    |
+| GetForecastByDateAsync | Yes  | No            | No     |
+
+## Location Source Availability
```
Comment (bderusha) [inline docs/data-source-matrix.md:47]:
```suggestion
## Location Coverage
```

**Comment by JenMadiedo** [comment]
@vaughanknight / @Willmish ready for review

File: docs/selecting-a-data-source.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+| Can Use Custom Data | &#10060;  | &#10060; | &#9989; |
+| Supports Trial + Full Account | &#9989;  | &#9989; (*[different config required](./configuration.md#electricitymaps-configuration)) | N/A |
+
+## Data Source Methods Available
+
+Not all data sources support all the routes provided in the interfaces (`IEmissionsDataSource`/`IForecastDataSource`).
+
+| Methods | WattTime  | ElectricityMaps | JSON | CLI Usage | Web Api Usage | SDK Usage
+|--------------|:-----------:|:-----------------:|:------:|:-:|:-:|:-:|
+| GetCarbonIntensityAsync | &#9989;  | &#9989; | &#9989; |`emissions`|`emissions/bylocation` or `emissions/bylocations` or `emissions/bylocations/best` or `emissions/average-carbon-intensity` or `average-carbon-intensity/batch`|`GetEmissionsDataAsync(...)` or `GetBestEmissionsDataAsync(...)` or `GetAverageCarbonIntensityDataAsync(...)`|
```
Comment (bderusha) [inline docs/selecting-a-data-source.md:28]: The forced wrapping on the hyphenated paths makes this table a bit tricky to read (IMO)

Consider swapping out the hyphen character for the [non-breaking hyphen](https://www.fileformat.info/info/unicode/char/2011/index.htm) `&#8209;` 

`<code>emissions/average&#8209;carbon&#8209;intensity</code>`

File: docs/selecting-a-data-source.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+| GetCurrentForecastAsync | &#9989;  | &#9989; | &#10060; |`emissions-forecasts`|`forecasts/current`|`GetCurrentForecastAsync(...)`|
```
Comment (bderusha) [inline docs/selecting-a-data-source.md:29]: `<code>emissions&#8209;forecasts</code>`

File: docs/selecting-a-data-source.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+| GetForecastByDateAsync | &#9989;  | &#10060; | &#10060; |`emissions-forecasts --requested-at`|`forecasts/batch` with `requestedAt` field|`GetForecastByDateAsync(...)`|
```
Comment (bderusha) [inline docs/selecting-a-data-source.md:30]: `<code>emissions&#8209;forecasts &#8209;&#8209;requested&#8209;at</code>`

**Review by bderusha — APPROVED** [review]
:shipit:

**Review by juzuluag — APPROVED** [review]
👍

**Review by vaughanknight — APPROVED** [review]
Nice update and thanks for this.  It's something that's come up a few times and we had discussed last week someone left a bug because they didn't know how to configure, when everything was actually working as intended... just not how they expected or were aware.  Big thanks.
